File: filtro_cinza.py
import sys
import time
import threading
from PIL import Image

def converter_para_escala_de_cinza(imagem, inicio: int, fim: int, altura: int):
    largura = fim - inicio
    nova_imagem = Image.new("RGB", (largura, altura))
    for x in range(largura):
        for y in range(altura):
            r, g, b = imagem.getpixel((inicio + x, y))
            valor_cinza = int(0.299 * r + 0.587 * g + 0.114 * b)
            nova_imagem.putpixel((x, y), (valor_cinza, valor_cinza, valor_cinza))
    return nova_imagem

def main():
    if len(sys.argv) != 3:
        print("Uso: python processador_sequencial.py <imagem_de_entrada> <imagem_de_saida>")
        sys.exit(1)

    caminho_entrada = sys.argv[1]
    caminho_saida = sys.argv[2]

    try:
        imagem_original = Image.open(caminho_entrada)
        if imagem_original.mode != 'RGB':
            raise Exception("Desculpe, imagem não está no formato RGB")
    except FileNotFoundError:
        print(f"Erro: O arquivo '{caminho_entrada}' não foi encontrado.")
        sys.exit(1)

    print(f"Processando a imagem '{caminho_entrada}' de forma sequencial...")
    
    inicio = time.time()

    imagem_processada = converter_para_escala_de_cinza(imagem_original, 0, imagem_original.width, imagem_original.height)
    
    fim = time.time()

    tempo_execucao_ms = (fim - inicio) * 1000
    imagem_processada.save(caminho_saida)
    print(f"Imagem convertida com sucesso e salva em '{caminho_saida}'")
    print(f"Tempo de execução (sequencial): {tempo_execucao_ms:.2f} ms\n")

    inicio2 = time.time()
    print(f"Processando a imagem '{caminho_entrada}' com multithreading...")

    num_threads = 6
    imagem_processada2 = Image.new("RGB", imagem_original.size)
    largura = imagem_original.width
    intervalo = largura // num_threads

    resultados = [None] * num_threads
    def worker(idx, inicio, fim):
        resultados[idx] = converter_para_escala_de_cinza(imagem_original, inicio, fim, imagem_original.height)

    threads = []
    for i in range(num_threads):
        inicio_parte = i * intervalo
        fim_parte = (i + 1) * intervalo if i < (num_threads - 1) else largura
        thread = threading.Thread(target=worker, args=(i, inicio_parte, fim_parte))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    for j in range(num_threads):
        imagem_processada2.paste(resultados[j], (j * intervalo, 0))

    fim2 = time.time()
    
    caminho_multithread = caminho_saida.replace(".jpg", "_multithread.jpg")
    imagem_processada2.save(caminho_multithread)
    
    tempo_execucao_ms2 = (fim2 - inicio2) * 1000
    
    print(f"Imagem convertida com sucesso e salva em '{caminho_multithread}'")
    print(f"Tempo de execução (sequencial): {tempo_execucao_ms2:.2f} ms")

if __name__ == "__main__":
    main()

File: filtro_cinza.java
import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.awt.Graphics;
import java.io.File;
import java.io.IOException;

public class filtro_cinza {

    // Função que converte parte da imagem para escala de cinza
    public static BufferedImage converterParaEscalaDeCinza(BufferedImage imagem, int inicio, int fim, int altura) {
        int largura = fim - inicio;
        BufferedImage novaImagem = new BufferedImage(largura, altura, BufferedImage.TYPE_INT_RGB);

        for (int x = 0; x < largura; x++) {
            for (int y = 0; y < altura; y++) {
                int rgb = imagem.getRGB(inicio + x, y);

                int r = (rgb >> 16) & 0xff;
                int g = (rgb >> 8) & 0xff;
                int b = rgb & 0xff;

                int valorCinza = (int)(0.299 * r + 0.587 * g + 0.114 * b);
                int novoRGB = (valorCinza << 16) | (valorCinza << 8) | valorCinza;

                novaImagem.setRGB(x, y, novoRGB);
            }
        }
        return novaImagem;
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.out.println("Uso: java ProcessadorImagem <imagem_de_entrada> <imagem_de_saida>");
            System.exit(1);
        }

        String caminhoEntrada = args[0];
        String caminhoSaida = args[1];

        BufferedImage imagemOriginal;
        try {
            imagemOriginal = ImageIO.read(new File(caminhoEntrada));
            if (imagemOriginal == null) {
                throw new IOException("Imagem não está no formato RGB");
            }
        } catch (IOException e) {
            System.out.println("Erro ao carregar a imagem: " + e.getMessage());
            return;
        }

        System.out.println("Processando a imagem '" + caminhoEntrada + "' de forma sequencial...");
        long inicio = System.currentTimeMillis();

        BufferedImage imagemProcessada = converterParaEscalaDeCinza(imagemOriginal, 0, imagemOriginal.getWidth(), imagemOriginal.getHeight());

        long fim = System.currentTimeMillis();
        long tempoExecucao = fim - inicio;

        try {
            ImageIO.write(imagemProcessada, "jpg", new File(caminhoSaida));
            System.out.println("Imagem convertida com sucesso e salva em '" + caminhoSaida + "'");
        } catch (IOException e) {
            System.out.println("Erro ao salvar a imagem: " + e.getMessage());
        }

        System.out.printf("Tempo de execução (sequencial): %.2f ms\n\n", (double)tempoExecucao);

        // ================= Multithreading =================
        System.out.println("Processando a imagem '" + caminhoEntrada + "' com multithreading...");

        int numThreads = 6;
        BufferedImage imagemProcessada2 = new BufferedImage(imagemOriginal.getWidth(), imagemOriginal.getHeight(), BufferedImage.TYPE_INT_RGB);
        int largura = imagemOriginal.getWidth();
        int intervalo = largura / numThreads;

        Thread[] threads = new Thread[numThreads];
        BufferedImage[] resultados = new BufferedImage[numThreads];

        inicio = System.currentTimeMillis();

        for (int i = 0; i < numThreads; i++) {
            final int idx = i;
            final int inicioParte = i * intervalo;
            final int fimParte = (i == numThreads - 1) ? largura : (i + 1) * intervalo;

            threads[i] = new Thread(() -> {
                resultados[idx] = converterParaEscalaDeCinza(imagemOriginal, inicioParte, fimParte, imagemOriginal.getHeight());
            });
            threads[i].start();
        }

        for (int i = 0; i < numThreads; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        // Junta os pedaços processados
        Graphics g = imagemProcessada2.getGraphics();
        for (int j = 0; j < numThreads; j++) {
            g.drawImage(resultados[j], j * intervalo, 0, null);
        }
        g.dispose();

        fim = System.currentTimeMillis();
        tempoExecucao = fim - inicio;

        String caminhoMultithread = caminhoSaida.replace(".jpg", "_multithread.jpg");
        try {
            ImageIO.write(imagemProcessada2, "jpg", new File(caminhoMultithread));
            System.out.println("Imagem convertida com sucesso e salva em '" + caminhoMultithread + "'");
        } catch (IOException e) {
            System.out.println("Erro ao salvar a imagem: " + e.getMessage());
        }

        System.out.printf("Tempo de execução (multithreading): %.2f ms\n", (double)tempoExecucao);
    }
}
